shared/dataflow/codeql/dataflow/VariableCapture.qll

```diff
@@ -0,0 +1,930 @@
+    string toString() {
+      exists(CapturedVariable v | this = TVariable(v) and result = v.toString())
+      or
+      result = "this" and this = TThis(_)
+    }
+  }
+
+  /** Holds if `cc` needs a definition at the entry of its callable scope. */
+  private predicate entryDef(CaptureContainer cc, BasicBlock bb, int i) {
+    exists(Callable c |
+      entryBlock(bb) and
+      pragma[only_bind_out](bb.getEnclosingCallable()) = c and
+      i =
+        -1 +
+          min(int j |
+            j = 1 or
+            captureRead(_, bb, j, _, _) or
+            captureWrite(_, bb, j, _, _) or
+            synthRead(_, bb, j, _, _)
+          )
+    |
+      cc = TThis(c)
+      or
+      exists(CapturedParameter p | cc = TVariable(p) and p.getCallable() = c)
+    )
+  }
+
+  private module CaptureSsaInput implements Ssa::InputSig {
+    class BasicBlock instanceof Input::BasicBlock {
+      string toString() { result = super.toString() }
+    }
+
+    BasicBlock getImmediateBasicBlockDominator(BasicBlock bb) {
+      result = Input::getImmediateBasicBlockDominator(bb)
+    }
+
+    BasicBlock getABasicBlockSuccessor(BasicBlock bb) {
+      result = Input::getABasicBlockSuccessor(bb)
+    }
+
+    class ExitBasicBlock extends BasicBlock {
+      ExitBasicBlock() { exitBlock(this) }
+    }
+
+    class SourceVariable = CaptureContainer;
+
+    predicate variableWrite(BasicBlock bb, int i, SourceVariable cc, boolean certain) {
+      (
+        exists(CapturedVariable v | cc = TVariable(v) and captureWrite(v, bb, i, true, _))
+        or
+        entryDef(cc, bb, i)
+      ) and
+      certain = true
+    }
+
+    predicate variableRead(BasicBlock bb, int i, SourceVariable cc, boolean certain) {
+      (
+        synthThisQualifier(bb, i) and cc = TThis(bb.(Input::BasicBlock).getEnclosingCallable())
+        or
+        exists(CapturedVariable v | cc = TVariable(v) |
+          captureRead(v, bb, i, true, _) or synthRead(v, bb, i, true, _)
+        )
+      ) and
+      certain = true
+    }
+  }
+
+  private module CaptureSsa = Ssa::Make<CaptureSsaInput>;
+
+  private newtype TClosureNode =
+    TSynthRead(CapturedVariable v, BasicBlock bb, int i, Boolean isPost) {
+      synthRead(v, bb, i, _, _)
+    } or
+    TSynthThisQualifier(BasicBlock bb, int i, Boolean isPost) { synthThisQualifier(bb, i) } or
+    TSynthPhi(CaptureSsa::DefinitionExt phi) {
+      phi instanceof CaptureSsa::PhiNode or phi instanceof CaptureSsa::PhiReadNode
+    } or
+    TExprNode(Expr expr, boolean isPost) {
+      expr instanceof VariableRead and isPost = [false, true]
+      or
+      exists(VariableWrite vw | expr = vw.getSource() and isPost = false)
+      or
+      synthRead(_, _, _, _, expr) and isPost = [false, true]
+    } or
+    TParamNode(CapturedParameter p) or
+    TThisParamNode(Callable c) { captureAccess(_, c) } or
+    TMallocNode(ClosureExpr ce) { hasConstructorCapture(ce, _) }
+
+  class ClosureNode extends TClosureNode {
```

Comment:
I think we should consider using `final` classes in all new code going forward. I.e.

```ql
private abstract class ClosureNodeImpl extends TClosureNode {
  abstract string toString();
  // ...
}

final class ClosureNode = ClosureNodeImpl;
```

Reply:
I agree that that's probably preferable, although I'm leaving this as-is for now.